### ISSUE TITLE
Installation failure; new breaking change in Python package

### ISSUE THREAD

**Issue by wbnns** [mail-in-a-box/mailinabox#979]
After successfully installing MIAB several times, including the latest version, this morning am noticing a failure on a fresh DigitalOcean instance (Ubuntu 14). Thinking maybe a Python package changed versions and caused a break on the MIAB side or something.

The following happens when running "curl -s https://mailinabox.email/setup.sh | sudo bash":

```
Running setup.py install for cryptography

    Installed /tmp/pip_build_root/cryptography/cffi-1.8.3-py3.4-linux-x86_64.egg
    Searching for pycparser
    Reading https://pypi.python.org/simple/pycparser/
    Best match: pycparser 2.15
    Downloading https://pypi.python.org/packages/37/05/65d342ba500c6f9dbb1978e9dd01a1097530e0291003638bbef329ad452$
/pycparser-2.15.tar.gz#md5=8629238be71db4d3f79b4fb1cd1397dc
    Processing pycparser-2.15.tar.gz
    Writing /tmp/easy_install-uy78zlxj/pycparser-2.15/setup.cfg
    Running pycparser-2.15/setup.py -q bdist_egg --dist-dir /tmp/easy_install-uy78zlxj/pycparser-2.15/egg-dist-tmp$
ui3jr2uq
    warning: no previously-included files matching 'yacctab.*' found under directory 'tests'
    warning: no previously-included files matching 'lextab.*' found under directory 'tests'
    warning: no previously-included files matching 'yacctab.*' found under directory 'examples'
    warning: no previously-included files matching 'lextab.*' found under directory 'examples'
    zip_safe flag not set; analyzing archive contents...
    pycparser.ply.__pycache__.yacc.cpython-34: module references __file__
    pycparser.ply.__pycache__.yacc.cpython-34: module MAY be using inspect.getsourcefile
    pycparser.ply.__pycache__.yacc.cpython-34: module MAY be using inspect.stack
    pycparser.ply.__pycache__.lex.cpython-34: module references __file__
    pycparser.ply.__pycache__.lex.cpython-34: module MAY be using inspect.getsourcefile
    pycparser.ply.__pycache__.ygen.cpython-34: module references __file__

    Installed /tmp/pip_build_root/cryptography/pycparser-2.15-py3.4.egg                                  
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip_build_root/cryptography/setup.py", line 334, in <module>
        **keywords_with_side_effects(sys.argv)
      File "/usr/lib/python3.4/distutils/core.py", line 108, in setup
        _setup_distribution = dist = klass(attrs)
      File "/usr/lib/python3/dist-packages/setuptools/dist.py", line 243, in __init__
        _Distribution.__init__(self,attrs)
      File "/usr/lib/python3.4/distutils/dist.py", line 280, in __init__
        self.finalize_options()
      File "/usr/lib/python3/dist-packages/setuptools/dist.py", line 277, in finalize_options
        ep.load()(self, ep.name, value)
      File "/tmp/pip_build_root/cryptography/cffi-1.8.3-py3.4-linux-x86_64.egg/cffi/setuptools_ext.py", line 181, i
n cffi_modules
        add_cffi_module(dist, cffi_module)
      File "/tmp/pip_build_root/cryptography/cffi-1.8.3-py3.4-linux-x86_64.egg/cffi/setuptools_ext.py", line 48, in
 add_cffi_module
        execfile(build_file_name, mod_vars)
      File "/tmp/pip_build_root/cryptography/cffi-1.8.3-py3.4-linux-x86_64.egg/cffi/setuptools_ext.py", line 24, in
 execfile
        exec(code, glob, glob)
      File "src/_cffi_src/build_openssl.py", line 81, in <module>
        extra_link_args=extra_link_args(compiler_type()),
      File "/tmp/pip_build_root/cryptography/src/_cffi_src/utils.py", line 61, in build_ffi_for_binding
        extra_link_args=extra_link_args,
      File "/tmp/pip_build_root/cryptography/src/_cffi_src/utils.py", line 70, in build_ffi
        ffi.cdef(cdef_source)
      File "/tmp/pip_build_root/cryptography/cffi-1.8.3-py3.4-linux-x86_64.egg/cffi/api.py", line 105, in cdef
        self._cdef(csource, override=override, packed=packed)
        File "/tmp/pip_build_root/cryptography/cffi-1.8.3-py3.4-linux-x86_64.egg/cffi/api.py", line 119, in[164/1087]
        self._parser.parse(csource, override=override, **options)
      File "/tmp/pip_build_root/cryptography/cffi-1.8.3-py3.4-linux-x86_64.egg/cffi/cparser.py", line 299, in parse
        self._internal_parse(csource)
      File "/tmp/pip_build_root/cryptography/cffi-1.8.3-py3.4-linux-x86_64.egg/cffi/cparser.py", line 304, in _inte
rnal_parse
        ast, macros, csource = self._parse(csource)
      File "/tmp/pip_build_root/cryptography/cffi-1.8.3-py3.4-linux-x86_64.egg/cffi/cparser.py", line 260, in _pars
e
        ast = _get_parser().parse(csource)
      File "/tmp/pip_build_root/cryptography/cffi-1.8.3-py3.4-linux-x86_64.egg/cffi/cparser.py", line 40, in _get_p
arser
        _parser_cache = pycparser.CParser()
      File "/tmp/pip_build_root/cryptography/pycparser-2.15-py3.4.egg/pycparser/c_parser.py", line 87, in __init__
        outputdir=taboutputdir)
      File "/tmp/pip_build_root/cryptography/pycparser-2.15-py3.4.egg/pycparser/c_lexer.py", line 66, in build
        self.lexer = lex.lex(object=self, **kwargs)
      File "/tmp/pip_build_root/cryptography/pycparser-2.15-py3.4.egg/pycparser/ply/lex.py", line 911, in lex
        lexobj.readtab(lextab, ldict)
      File "/tmp/pip_build_root/cryptography/pycparser-2.15-py3.4.egg/pycparser/ply/lex.py", line 233, in readtab
        titem.append((re.compile(pat, lextab._lexreflags | re.VERBOSE), _names_to_funcs(func_name, fdict)))
      File "/usr/lib/python3.4/re.py", line 223, in compile
        return _compile(pattern, flags)
      File "/usr/lib/python3.4/re.py", line 294, in _compile
        p = sre_compile.compile(pattern, flags)
      File "/usr/lib/python3.4/sre_compile.py", line 579, in compile
        "sorry, but this version only supports 100 named groups"
    AssertionError: sorry, but this version only supports 100 named groups
    Complete output from command /usr/bin/python3 -c "import setuptools, tokenize;__file__='/tmp/pip_build_root/cry
ptography/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__,
'exec'))" install --record /tmp/pip-n61uqnul-record/install-record.txt --single-version-externally-managed --compil
e:


Installed /tmp/pip_build_root/cryptography/cffi-1.8.3-py3.4-linux-x86_64.egg

Searching for pycparser

Reading https://pypi.python.org/simple/pycparser/

Best match: pycparser 2.15

Downloading https://pypi.python.org/packages/37/05/65d342ba500c6f9dbb1978e9dd01a1097530e0291003638bbef329ad4522/pyc
parser-2.15.tar.gz#md5=8629238be71db4d3f79b4fb1cd1397dc

Processing pycparser-2.15.tar.gz

Writing /tmp/easy_install-uy78zlxj/pycparser-2.15/setup.cfg

Running pycparser-2.15/setup.py -q bdist_egg --dist-dir /tmp/easy_install-uy78zlxj/pycparser-2.15/egg-dist-tmp-ui3j
r2uq

warning: no previously-included files matching 'yacctab.*' found under directory 'tests'

warning: no previously-included files matching 'lextab.*' found under directory 'tests'

warning: no previously-included files matching 'yacctab.*' found under directory 'examples'

warning: no previously-included files matching 'lextab.*' found under directory 'examples'

zip_safe flag not set; analyzing archive contents...

pycparser.ply.__pycache__.yacc.cpython-34: module references __file__

pycparser.ply.__pycache__.yacc.cpython-34: module MAY be using inspect.getsourcefile

pycparser.ply.__pycache__.yacc.cpython-34: module MAY be using inspect.stack

pycparser.ply.__pycache__.lex.cpython-34: module references __file__

pycparser.ply.__pycache__.lex.cpython-34: module MAY be using inspect.getsourcefile

pycparser.ply.__pycache__.ygen.cpython-34: module references __file__



Installed /tmp/pip_build_root/cryptography/pycparser-2.15-py3.4.egg

Traceback (most recent call last):

  File "<string>", line 1, in <module>

  File "/tmp/pip_build_root/cryptography/setup.py", line 334, in <module>

   **keywords_with_side_effects(sys.argv)

  File "/usr/lib/python3.4/distutils/core.py", line 108, in setup

    _setup_distribution = dist = klass(attrs)

  File "/usr/lib/python3/dist-packages/setuptools/dist.py", line 243, in __init__

    _Distribution.__init__(self,attrs)

  File "/usr/lib/python3.4/distutils/dist.py", line 280, in __init__

    self.finalize_options()

  File "/usr/lib/python3/dist-packages/setuptools/dist.py", line 277, in finalize_options

    ep.load()(self, ep.name, value)

  File "/tmp/pip_build_root/cryptography/cffi-1.8.3-py3.4-linux-x86_64.egg/cffi/setuptools_ext.py", line 181, in cf
fi_modules

    add_cffi_module(dist, cffi_module)

  File "/tmp/pip_build_root/cryptography/cffi-1.8.3-py3.4-linux-x86_64.egg/cffi/setuptools_ext.py", line 48, in add
_cffi_module

    execfile(build_file_name, mod_vars)

  File "/tmp/pip_build_root/cryptography/cffi-1.8.3-py3.4-linux-x86_64.egg/cffi/setuptools_ext.py", line 24, in ex$
cfile

    exec(code, glob, glob)

  File "src/_cffi_src/build_openssl.py", line 81, in <module>

    extra_link_args=extra_link_args(compiler_type()),

  File "/tmp/pip_build_root/cryptography/src/_cffi_src/utils.py", line 61, in build_ffi_for_binding

    extra_link_args=extra_link_args,

  File "/tmp/pip_build_root/cryptography/src/_cffi_src/utils.py", line 70, in build_ffi

    ffi.cdef(cdef_source)

  File "/tmp/pip_build_root/cryptography/cffi-1.8.3-py3.4-linux-x86_64.egg/cffi/api.py", line 105, in cdef

    self._cdef(csource, override=override, packed=packed)

  File "/tmp/pip_build_root/cryptography/cffi-1.8.3-py3.4-linux-x86_64.egg/cffi/api.py", line 119, in _cdef


    self._parser.parse(csource, override=override, **options)

  File "/tmp/pip_build_root/cryptography/cffi-1.8.3-py3.4-linux-x86_64.egg/cffi/cparser.py", line 299, in parse

    self._internal_parse(csource)

  File "/tmp/pip_build_root/cryptography/cffi-1.8.3-py3.4-linux-x86_64.egg/cffi/cparser.py", line 304, in _internal
_parse

    ast, macros, csource = self._parse(csource)

  File "/tmp/pip_build_root/cryptography/cffi-1.8.3-py3.4-linux-x86_64.egg/cffi/cparser.py", line 260, in _parse

    ast = _get_parser().parse(csource)

  File "/tmp/pip_build_root/cryptography/cffi-1.8.3-py3.4-linux-x86_64.egg/cffi/cparser.py", line 40, in _get_parse
r

    _parser_cache = pycparser.CParser()

  File "/tmp/pip_build_root/cryptography/pycparser-2.15-py3.4.egg/pycparser/c_parser.py", line 87, in __init__

    outputdir=taboutputdir)

  File "/tmp/pip_build_root/cryptography/pycparser-2.15-py3.4.egg/pycparser/c_lexer.py", line 66, in build

    self.lexer = lex.lex(object=self, **kwargs)

                                                                                                         [2/1087]
  File "/tmp/pip_build_root/cryptography/pycparser-2.15-py3.4.egg/pycparser/ply/lex.py", line 911, in lex

    lexobj.readtab(lextab, ldict)

  File "/tmp/pip_build_root/cryptography/pycparser-2.15-py3.4.egg/pycparser/ply/lex.py", line 233, in readtab

    titem.append((re.compile(pat, lextab._lexreflags | re.VERBOSE), _names_to_funcs(func_name, fdict)))

  File "/usr/lib/python3.4/re.py", line 223, in compile

    return _compile(pattern, flags)

  File "/usr/lib/python3.4/re.py", line 294, in _compile

    p = sre_compile.compile(pattern, flags)

  File "/usr/lib/python3.4/sre_compile.py", line 579, in compile

    "sorry, but this version only supports 100 named groups"

AssertionError: sorry, but this version only supports 100 named groups

----------------------------------------
Cleaning up...
Command /usr/bin/python3 -c "import setuptools, tokenize;__file__='/tmp/pip_build_root/cryptography/setup.py';exec(
compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --recor
d /tmp/pip-n61uqnul-record/install-record.txt --single-version-externally-managed --compile failed with error code
1 in /tmp/pip_build_root/cryptography
Storing debug log for failure in /root/.pip/pip.log
-----------------------------------------
```

**Comment by wbinns** [comment]
@yodax Hey, just a quick update on this. I started over from scratch and did not run `apt-get update && apt-get upgrade` before doing the installation. I completed the installation successfully when I didn't do that.

**Comment by yodax** [comment]
Cool, I was just trying to reproduce this on a fresh box.

**Comment by yodax** [comment]
Fresh install works for me as well. I blame "Solar Flares"

**Comment by bronson** [comment]
Sounds like a bug in PyCParser 2.14: http://stackoverflow.com/questions/39829473/cryptography-assertionerror-sorry-but-this-version-only-supports-100-named-gro

Downgrading to 2.13 is the popular fix.  Installing `--no-binary` is also reported to work.

**Comment by wbinns** [comment]
@yodax @bronson Thanks guys! Closing this issue for now, then. 👍

**Comment by anoma** [comment]
This should be reopened as currently it is causing new "mail in a box" installs to completely fail.

temp fix

`pip install git+https://github.com/eliben/pycparser@release_v2.14`

**Comment by bronson** [comment]
anoma, is that the right version?  I thought v2.14 was the broken one.

**Comment by anoma** [comment]
TBH i dont know what the perfect version is but I can confirm that is the command i used to bypass the error and complete the install successfully.

**Comment by JoshData** [comment]
It works for me on a fresh install within Vagrant.

**Comment by anoma** [comment]
If someone can tell me how to revert the temp fix above I will try another install run

**Comment by JoshData** [comment]
You could try `pip3 uninstall pycparser`.

**Comment by anoma** [comment]
Confused.

After `pip3 uninstall pycparser` failed I noted that I had used `pip` earlier. 

I used  `pip uninstall pycparser` successfully and reinstalled mailinabox.

I ended up with:

```
root@box:~# pip3 show pycparser
---
Name: pycparser
Version: 2.17
Location: /usr/local/lib/python3.4/dist-packages
Requires:
```

```
root@box:~# pip show pycparser
root@box:~# 
```

How can this module dependency no longer be required? Was it a stock python2 install that we dont actually need that just happened to be broken and this failed the install procedure?
